crates/wasi/src/preview2/tcp.rs

```diff
@@ -0,0 +1,284 @@
+    /// The part of a `HostTcpSocket` which is reference-counted so that we
+    /// can pass it to async tasks.
+    pub(crate) inner: Arc<HostTcpSocketInner>,
+}
+
+/// The inner reference-counted state of a `HostTcpSocket`.
+pub(crate) struct HostTcpSocketInner {
+    /// On Unix-family platforms we can use `AsyncFd` for efficient polling.
+    #[cfg(unix)]
+    pub(crate) tcp_socket: tokio::io::unix::AsyncFd<cap_std::net::TcpListener>,
+
+    /// On non-Unix, we can use plain `poll`.
+    #[cfg(not(unix))]
+    pub(crate) tcp_socket: cap_std::net::TcpListener,
+
+    /// The current state in the bind/listen/accept/connect progression.
+    pub(crate) tcp_state: RwLock<HostTcpState>,
+}
+
+impl HostTcpSocket {
+    /// Create a new socket in the given family.
+    pub fn new(family: AddressFamily) -> io::Result<Self> {
+        // Create a new host socket and set it to non-blocking, which is needed
+        // by our async implementation.
+        let tcp_socket = TcpListener::new(family, Blocking::No)?;
+
+        // On Unix, pack it up in an `AsyncFd` so we can efficiently poll it.
+        #[cfg(unix)]
+        let tcp_socket = tokio::io::unix::AsyncFd::new(tcp_socket)?;
+
+        Ok(Self {
+            inner: Arc::new(HostTcpSocketInner {
+                tcp_socket,
+                tcp_state: RwLock::new(HostTcpState::Default),
+            }),
+        })
+    }
+
+    /// Create a `HostTcpSocket` from an existing socket.
+    ///
+    /// The socket must be in non-blocking mode.
+    pub fn from_tcp_stream(tcp_socket: cap_std::net::TcpStream) -> io::Result<Self> {
+        let fd = rustix::fd::OwnedFd::from(tcp_socket);
+        let tcp_socket = TcpListener::from(fd);
+
+        // On Unix, pack it up in an `AsyncFd` so we can efficiently poll it.
+        #[cfg(unix)]
+        let tcp_socket = tokio::io::unix::AsyncFd::new(tcp_socket)?;
+
+        Ok(Self {
+            inner: Arc::new(HostTcpSocketInner {
+                tcp_socket,
+                tcp_state: RwLock::new(HostTcpState::Default),
+            }),
+        })
+    }
+
+    pub fn tcp_socket(&self) -> &cap_std::net::TcpListener {
+        self.inner.tcp_socket()
+    }
+
+    pub fn clone_inner(&self) -> Arc<HostTcpSocketInner> {
+        Arc::clone(&self.inner)
+    }
+
+    /// Acquire a reader lock for `self.tcp_state`.
+    pub fn tcp_state_read_lock(&self) -> RwLockReadGuard<HostTcpState> {
+        self.inner.tcp_state.read().unwrap()
+    }
+
+    /// Acquire a writer lock for `self.tcp_state`.
+    pub fn tcp_state_write_lock(&self) -> RwLockWriteGuard<HostTcpState> {
+        self.inner.tcp_state.write().unwrap()
+    }
+}
+
+impl HostTcpSocketInner {
+    pub fn tcp_socket(&self) -> &cap_std::net::TcpListener {
+        let tcp_socket = &self.tcp_socket;
+
+        // Unpack the `AsyncFd`.
+        #[cfg(unix)]
+        let tcp_socket = tcp_socket.get_ref();
+
+        tcp_socket
+    }
+
+    pub fn set_state(&self, new_state: HostTcpState) {
```

Comment:
from my reading of these changes, `set_state` is only ever performed when in an `&mut self` method of `WasiView`, rather than in a spawned task, which makes me think theres no longer a need for an arc and locking for that member. Can we lift it out to be in HostTcpSocket and manipulate it with plain old &mut?

Reply:
Done